Add Tracks Analytics on frontend

mehmoodak opened the pull request:
### Description

Add Tracks Analytics by using the `@automattic/calypso-analytics` package. With each track event, the following properties will be tracked:

- Properties related to VIP Platform (e.g. org_id, env_id).
- Properties related to Remote Data Blocks (e.g. plugin_version).
- Event specific properties.

**Note:** This tracking will only work if site is hosted on VIP Platform or if the analytics is enabled via filter.

### How to test?

1. Create a site using `vip dev-env` which will have the MU Plugins.
1. Move `remote-data-blocks` plugin to the `plugins` folder.
1. Add filter in `plugin-loader.php` to enable tracking.

    ```php
    add_filter( 'remote_data_blocks_enable_tracks_analytics', '__return_true' );
    ```
1. Checkout [this wp-calypso PR](https://github.com/Automattic/wp-calypso/pull/96007), `build` and `link` the `@automattic/calypso-analytics` package.
1. To allow tracking on local env comment check on [PHP](https://github.com/Automattic/remote-data-blocks/blob/1574d6f729327eda7a10ecfa686c58a3252f8187/inc/Analytics/TracksAnalytics.php#L30C1-L34C4) and [JS](https://github.com/Automattic/remote-data-blocks/blob/1574d6f729327eda7a10ecfa686c58a3252f8187/src/blocks/remote-data-container/utils/tracks.ts#L15C1-L19C3) side.
1. Trigger track events, use "Live View" to verify that the events are getting logged.

mehmoodak committed:
➕ ADDS: function to check if analytics is enabled

mehmoodak committed:
➕ ADDS: track_event function

mehmoodak committed:
➕ ADDS: events for activation and deactivation

mehmoodak committed:
➕ ADDS: remotedatablocks_data_sources_stats event

mehmoodak committed:
➕ ADDS: remotedatablocks_data_sources events

mehmoodak committed:
🐛 FIX: track prefix

mehmoodak committed:
🔨 REFACTOR: change event name

mehmoodak committed:
💫 UPDATE: remotedatablocks_view_data_sources event to only track once a day

mehmoodak committed:
💫 UPDATE: event name

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
➕ ADDS: remotedatablocks_usage_stats event

mehmoodak committed:
🔨 REFACTOR: move hooks tracking to Analytics class

mehmoodak committed:
🔨 REFACTOR: class name from Tracks to TracksAnalytics

mehmoodak committed:
🔨 REFACTOR: remove is_enabled function from TracksAnalytics

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
➕ ADDS: tests for get_service method

mehmoodak committed:
➕ ADDS: tests for get_data_source method

mehmoodak committed:
move tracks code to functions for easy mocking

mehmoodak committed:
update property $tracks to $instance

mehmoodak committed:
update TracksAnalytics to remove static properties and functions

mehmoodak committed:
➕ ADDS: tests for TracksAnalyticsTest

mehmoodak committed:
➕ ADDS: few tests for TracksAnalytics

mehmoodak committed:
🐛 FIX: tests

mehmoodak committed:
🐛 FIX: psalm error

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
➕ ADDS: tests for TracksAnalyticsTest

mehmoodak committed:
➖ REMOVE: full stop from comments

mehmoodak committed:
👌 IMPROVE: comments

mehmoodak committed:
💫 UPDATE: methods to static

mehmoodak committed:
Remove duplicate from psalm.xml

Co-authored-by: Chris Zarate <chris.zarate@automattic.com>

mehmoodak committed:
Use nullish coalescing operator

Co-authored-by: Chris Zarate <chris.zarate@automattic.com>

chriszarate committed:
Add apply filters global

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
💫 UPDATE: tests to work properly with static methods

mehmoodak committed:
ADDS: @automattic/calypso-analytics package

mehmoodak committed:
sendTracksEvent util function

mehmoodak committed:
➕ ADDS: track events for Field Shortcode

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
Merge branch 'cafe-942/add-tracks-analytics' into cafe-942/add-tracks-analytics-on-js

mehmoodak committed:
💫 UPDATE: sendTracksEvent to not track on local

mehmoodak committed:
➕ ADDS: track actions on remote data container

mehmoodak committed:
➕ ADDS: track actions for block binding controls

mehmoodak committed:
➕ ADDS: data source in REMOTE_DATA_BLOCKS

mehmoodak committed:
➕ ADDS: track events while adding block

mehmoodak committed:
➕ ADDS: track event for override

mehmoodak committed:
➕ ADDS: track event while associating pattern to block type

mehmoodak committed:
➕ ADDS: base props for tracks

mehmoodak committed:
➕ ADDS: avoid tracking events on local env

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
Merge branch cafe-942/add-tracks-analytics into cafe-942/add-tracks-analytics-on-js

mehmoodak committed:
🐛 FIX: linting

mehmoodak committed:
💫 UPDATE: variable name

mehmoodak committed:
👌 IMPROVE: comments

mehmoodak committed:
👌 IMPROVE: naming

mehmoodak committed:
👌 IMPROVE: additional track props on DataSourceController

mehmoodak committed:
🐛 FIX: tracking of plugin activation and deactivation

mehmoodak committed:
Merge branch cafe-942/add-tracks-analytics into cafe-942/add-tracks-analytics-on-js

mehmoodak committed:
Add tests for tracks.ts

mehmoodak committed:
fix types

mehmoodak committed:
👌 IMPROVE: RDB tracks props

mehmoodak committed:
💫 UPDATE: remove props that are being sent by MU Plugins

mehmoodak committed:
Merge branch 'cafe-942/add-tracks-analytics' into cafe-942/add-tracks-analytics-on-js

mehmoodak committed:
➖ REMOVE: extra data sources from interface and props

mehmoodak committed:
👌 IMPROVE: naming

mehmoodak committed:
➖ REMOVE: extra spacing

mehmoodak committed:
➕ ADDS: tests for remotedatablocks_add_block event

mehmoodak committed:
🐛 FIX: empty data source and improve action names

mehmoodak committed:
track event prop type

chriszarate commented on types/globals.d.ts:
```diff
@@ -1,6 +1,22 @@
+	var VIP_TRACKS_BASE_PROPS: VipTracksBaseProps | undefined;
+}
+
+export interface VipTracksBaseProps {
```
Put this in a separate file like `tracks.d.ts`

mehmoodak replied:
https://github.com/Automattic/remote-data-blocks/pull/185/commits/be36ae58a7d2ef1d97a1d4591a620de7ac137190

chriszarate commented on src/blocks/remote-data-container/utils/tracks.ts:
```diff
@@ -0,0 +1,26 @@
+ * Send a tracks event with the given name and properties.
+ */
+export function sendTracksEvent( eventName: string, props: Record< string, unknown > ): void {
+	const vipBaseProps = window.VIP_TRACKS_BASE_PROPS;
```
Is a global necessary? Can the Tracks library expose a way to register these base props?

mehmoodak replied:
This global is used to pass information from PHP to the frontend. I did some search about a better way to pass this information but couldn't find anything.

chriszarate replied:
I understand, I just don't think mu-plugins should be responsible for this

mehmoodak replied:
Agreed. The main reason of passing this information from MU Plugins is to have a consistent implementation across all of VIP.

> Can the Tracks library expose a way to register these base props?

By registering you mean?

chriszarate replied:
> The main reason of passing this information from MU Plugins is to have a consistent implementation across all of VIP.

I am only objecting to the JS global as the method of passing the base props. Currently we are injecting it into the `<head>` which is not ideal. Providing it as a global also provides little value, since each plugin still needs custom code (and types) to access the global and inject it. I would rather avoid the JS global and script in `<head>`, have the plugins load the base props via PHP, and then inject it into their own global (e.g., `REMOTE_DATA_BLOCKS_SETTINGS` for this plugin). Alternatively plugins could integrate via a REST endpoint or a custom library.

> By registering you mean?

Another approach would be to render the base props as meta tag(s) and have the client-side code read them (once) and include them on each event.

mehmoodak replied:
> and then inject it into their own global (e.g., REMOTE_DATA_BLOCKS_SETTINGS for this plugin)

I will implement it in `REMOTE_DATA_BLOCKS_SETTINGS` const.

> Alternatively plugins could integrate via a REST endpoint or a custom library.

I have initially suggested it in P2 and Mo feedback on this is that using the REST API is more expensive then making the calls directly to Tracks from the client.

> Another approach would be to render the base props as meta tag(s) and have the client-side code read them (once) and include them on each event.

It's more parsing, which I think is better to avoid.

mehmoodak replied:
> I will implement it in `REMOTE_DATA_BLOCKS` const.

Done. https://github.com/Automattic/remote-data-blocks/pull/185/commits/44a3a22af29ccd0e0eead8489809e8bfc9667723

chriszarate commented on .husky/pre-commit:
```diff
@@ -1,6 +1,7 @@
 fi
 
 lint-staged
+npm run check-types # Run tsc on the whole project, as running it only on staged files may result in import or type errors.
```
Probably better to think of type checking as something different from "linting" and remove it from scripts with that name.

mehmoodak replied:
https://github.com/Automattic/remote-data-blocks/pull/185/commits/7ce75c6689953bc0eb331798e4bd2bc6a7c39717

chriszarate commented on inc/Editor/BlockManagement/BlockRegistration.php:
```diff
@@ -77,6 +78,7 @@ public static function register_blocks(): void {
 				'availableBindings' => $available_bindings,
 				'loop' => $config['loop'],
 				'name' => $block_name,
+				'dataSource' => ConfigStore::get_data_source( $block_name ),
```
Is this trying to describe the type of data source, like Airtable or Shopify? Maybe we can name it `dataSourceType`

mehmoodak replied:
https://github.com/Automattic/remote-data-blocks/pull/185/commits/a3e42f9527602ce6ac3f10a2fb44e0ea57e4c442

chriszarate commented on inc/REST/RemoteDataController.php:
```diff
@@ -77,6 +77,7 @@ public static function execute_query( WP_REST_Request $request ): array|WP_Error
 		return array_merge(
 			[
 				'block_name' => $block_name,
+				'data_source' => ConfigStore::get_data_source( $block_name ),
```
same here

mehmoodak replied:
Removed it. I was previously passing the `data_source` prop, but now I'm using the block name to retrieve the data source instead.

chriszarate commented on src/blocks/remote-data-container/components/BlockBindingControls.tsx:
```diff
@@ -68,6 +84,11 @@ export function BlockBindingControls( props: BlockBindingControlsProps ) {
 			: undefined;
 		updateBinding( 'content', { ...contentArgs, field: contentField, label } );
+		sendTracksEvent( 'remotedatablocks_remote_data_container_actions', {
+			action: 'show_label',
```
`toggle_label`?

chriszarate replied:
or `action: showLabel ? 'show_label' : 'hide_label'`

mehmoodak replied:
https://github.com/Automattic/remote-data-blocks/pull/185/commits/bf3289a1e22f6cff137c9504d48c7d29e8a877d9

chriszarate commented on src/blocks/remote-data-container/components/BlockBindingControls.tsx:
```diff
@@ -68,6 +84,11 @@ export function BlockBindingControls( props: BlockBindingControlsProps ) {
+		sendTracksEvent( 'remotedatablocks_remote_data_container_actions', {
+			action: 'show_label',
+			data_source: getBlockDataSource( remoteDataName ),
+			value: showLabel,
```
could probably be boolean

chriszarate commented on src/blocks/remote-data-container/components/field-shortcode/FieldShortcodeButton.tsx:
```diff
@@ -76,11 +79,20 @@ export function FieldShortcodeButton( props: WPFormatEditProps ) {
 		onClose();
+		sendTracksEvent( 'remotedatablocks_field_shortcode', {
+			action: data.action,
+			data_source: getBlockDataSource( data.remoteData?.blockName ),
```
seems like looking up the data source type should be delegated to `sendTracksEvent`

mehmoodak replied:
We don't have the `data_source` prop on all events, so I didn't delegate it to `sendTracksEvent`.

chriszarate replied:
Can it just be undefined in those cases?

mehmoodak replied:
It can be `undefined`, but we still need to pass `blockName`, so I think it's better to pass `data_source`, which is the value we ultimately send to `Tracks` anyway.

By the way, I've updated the types for these Tracks Events. (https://github.com/Automattic/remote-data-blocks/pull/185/commits/a14dfce570d4075348f37c5246b26fc89b93785c)

chriszarate commented on src/blocks/remote-data-container/components/modals/InputModal.tsx:
```diff
@@ -25,9 +28,14 @@ export function InputModal( props: InputModalProps ) {
 	}
 
-	function wrappedOnSelect(): void {
+	function onSaveClick(): void {
```
```suggestion
	function onSelectItem(): void {
```

mehmoodak replied:
https://github.com/Automattic/remote-data-blocks/pull/185/commits/cd8bfc1a8d90312e75aa033dc1dd8cc20a08084c

chriszarate commented on src/blocks/remote-data-container/filters/withBlockBinding.tsx:
```diff
@@ -66,6 +66,7 @@ function BoundBlockEdit( props: BoundBlockEditProps ) {
 						attributes={ attributes }
 						availableBindings={ availableBindings }
 						blockName={ blockName }
+						remoteDataName={ remoteDataName }
```
Instead of prop drilling the container block name, I think we should just prop drill the `sendTracksEvent` function with the data_source_type baked in

mehmoodak replied:
We'll need prop drilling in either case, to me passing `blockName` seems simpler than passing the function.

mehmoodak committed:
🔨 REFACTOR: move VipTracksBaseProps to tracks.d.ts

mehmoodak committed:
🔨 REFACTOR: rename get_data_source to get_data_source_type

mehmoodak committed:
🔨 REFACTOR: update action name

mehmoodak committed:
👌 IMPROVE: types of tracks events

mehmoodak committed:
🔨 REFACTOR: rename onSaveClick function to onSelectItem

mehmoodak committed:
🔨 REFACTOR: rename tracks event data_source property to data_source_type

mehmoodak committed:
➖ REMOVE: type checking from linting

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
Refactor EnvironmentConfig and TracksAnalytics classes

These changes improve the overall structure and maintainability of the codebase.

mehmoodak committed:
Merge branch 'cafe-942/add-tracks-analytics' into cafe-942/add-tracks-analytics-on-js

mehmoodak committed:
🐛 FIX: undefined wp function errors

mehmoodak committed:
Merge branch 'cafe-942/add-tracks-analytics' into cafe-942/add-tracks-analytics-on-js

mehmoodak committed:
🔨 REFACTOR: For Tracks remove VIP_TRACKS_BASE_PROPS const and use the REMOTE_DATA_BLOCKS const

mehmoodak committed:
🐛 FIX: ignore psalm lint error

mehmoodak committed:
Merge branch 'cafe-942/add-tracks-analytics' into cafe-942/add-tracks-analytics-on-js

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics

mehmoodak committed:
Merge branch 'cafe-942/add-tracks-analytics' into cafe-942/add-tracks-analytics-on-js

maxschmeling commented on tsconfig.json:
```diff
@@ -18,5 +18,5 @@
     "typeRoots": [ "./types", "./node_modules/@types" ]
   },
-  "include": [ "./types/*.d.ts", "./example/**/*", "./src/**/*", "./tests/**/*" ]
+  "include": [ "./**/*.d.ts", "./example/**/*", "./src/**/*", "./tests/**/*" ]
```
Was this needed for something specific?

mehmoodak replied:
Reverted. I was trying to fix some TS issue but this change is not needed now.

maxschmeling reviewed (approved):
Looks like a really good start. Some minor pedantic things I might do a little different, but  not worth nit picking. I like it.

maxschmeling commented on src/blocks/remote-data-container/components/BlockBindingControls.tsx:
```diff
@@ -51,11 +55,23 @@ export function BlockBindingControls( props: BlockBindingControlsProps ) {
 			removeBinding( target );
+			sendTracksEvent( 'remotedatablocks_remote_data_container_actions', {
+				action: 'remove_binding',
+				data_source_type: getBlockDataSourceType( remoteDataName ),
```
Should this also have a `remote_data_field` property?

mehmoodak replied:
For `remove_binding` action we don't get `field` value and that's why I haven't included `remote_data_field` in track props. Let me know if you think we should include it, I will find an alternative.

mehmoodak committed:
Merge branch 'trunk' into cafe-942/add-tracks-analytics-on-js

mehmoodak committed:
💫 UPDATE: revert tsconfig change
